Run MockMediaSource in the GPU Process

jyavenard opened the pull request:
#### 9faed3b669e65d2fcb6157b6aa8c483500d22547
<pre>
Run MockMediaSource in the GPU Process
<a href="https://bugs.webkit.org/show_bug.cgi?id=254074">https://bugs.webkit.org/show_bug.cgi?id=254074</a>
rdar://106858091

Reviewed by Youenn Fablet.

Related changes:
- Fix crashes when running/media/media-source/media-source-minimumupcomingpresentationtime.html,
some missing methods with the RemoteSourceBufferProxy required to run this
test.
- Use C++11 initializers in MockMediaPlayerMediaSource

Covered by existing tests.

* LayoutTests/media/media-source/media-source-minimumupcomingpresentationtime.html: whitespace changes.
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::resetMediaEngines): Deleted.
* Source/WebCore/html/HTMLMediaElement.h: Delete resetMediaEngines, use mediaStragegy&apos;s resetMediaEngines instead.
* Source/WebCore/page/DeprecatedGlobalSettings.cpp: Use resetMediaEngines from mediaStrategy.
(WebCore::DeprecatedGlobalSettings::setAVFoundationEnabled):
(WebCore::DeprecatedGlobalSettings::setGStreamerEnabled):
* Source/WebCore/platform/MediaStrategy.cpp:
(WebCore::MediaStrategy::resetMediaEngines):
(WebCore::MediaStrategy::enableMockMediaSource):
(WebCore::MediaStrategy::mockMediaSourceEnabled const):
(WebCore::MediaStrategy::addMockMediaSourceEngine):
* Source/WebCore/platform/MediaStrategy.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(buildMediaEnginesVector()): Add MockMSE support to registerRemoteEngine.
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp: Use C++11 initializers.
(WebCore::MockMediaPlayerMediaSource::MockMediaPlayerMediaSource):
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::initializeMockMediaSource):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::enableMockMediaSource):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::updateCachedState):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::minimumUpcomingPresentationTimeForTrackID):
(WebKit::RemoteSourceBufferProxy::setMaximumQueueDepthForTrackID):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::load):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::minimumUpcomingPresentationTimeForTrackID):
(WebKit::SourceBufferPrivateRemote::setMaximumQueueDepthForTrackID):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::enableMockMediaSource):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:
(WebKit::WebMediaStrategy::enableMockMediaSource): Add method

Canonical link: <a href="https://commits.webkit.org/262173@main">https://commits.webkit.org/262173@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1047a724c63a882ad89b9e6f6695e3555dd59a4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/758 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/780 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/807 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1006 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/677 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/749 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/829 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/863 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/903 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/765 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/730 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/721 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/955 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/714 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/731 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/690 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/736 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1703 "260 api tests failed or timed out") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/721 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/703 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/674 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/721 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/189 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/731 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3888,6 +3889,13 @@ void Page::abortApplePayAMSUISession(ApplePayAMSUIPaymentHandler& paymentHandler
 #endif // ENABLE(APPLE_PAY_AMS_UI)
 
+#if ENABLE(MEDIA_SOURCE)
+void Page::setMockMediaSourceEnabled()
```
I would name it enableMockMediaSource() given it has no boolean parameter

jyavenard replied:
ok

youennf commented on Source/WebCore/platform/MediaStrategy.cpp:
```diff
@@ -37,4 +42,28 @@ std::unique_ptr<NowPlayingManager> MediaStrategy::createNowPlayingManager() cons
+    return false;
+};
+
+void MediaStrategy::enableMockMediaSource()
```
If we update setMockMediaSourceEnabled to enableMockMediaSource, we would need to rename this one to something like addMockMediaSourceEngine

jyavenard replied:
ok

youennf commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -4437,13 +4433,10 @@ bool Internals::isPluginSnapshotted(Element&)
+    Document* document = contextDocument();
+    if (!document || !document->page())
+        return;
+    document->page()->setMockMediaSourceEnabled();
```
If we add it there, we reuse the same GPUProcess and WebProcess for other tests, so the mock might be enabled even though it was not supposed to be enabled.
This is a pre-existing issue so maybe this is fine.
If we want to fix this issue, one possibility would be to move this API to test runner.
Then test runner would kill the GPUProcess when resetting the test options for the next test if mock MSE is enabled.

jyavenard replied:
It will get reset upon the next run as enabling the MockMSE in the GPU Process doesn't really impact what is happening in the content process.
Setting the pref that enable or disable the AVFoundation MediaPlayer will automatically reset the available MediaPlayer list in the content process.

With the RemoteMediaPlayer, the content process sends which MediaPlayer to use to the GPU Process. If it's disabled in the current content process, that's where the filtering is happening and as such attempting to open a RemoteMediaPlayer will fail.

I can run a call to disable it in the GPU Process if need be, but in practice it makes no difference.

jyavenard replied:
on 2nd thought, it we enable the MockMediaSource in the GPU process by adding the MockMediaPlayerMediaSource to the list of supported MediaPlayer, and we expect tests and content processes to be running concurrently, we don't want to do what we are doing in the CP when enabling the MockMediaSource: which is to reset the list of supported MediaPlayers and only add the MockMediaPlayerMediaSource ; We only want to to add it to the existing supported list.

Otherwise from that point on, only MockMediaSource is usable, in all content process regardless of what the test is doing.

Considering what I wrote above (that only what is allowed in the CP matters) I'll remove all code that would otherwise reset the MediaPlayer list.

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
```diff
@@ -957,8 +957,10 @@ void RemoteMediaPlayerProxy::updateCachedState(bool forceCurrentTimeUpdate)
         m_bufferedChanged = false;
-        if (m_engineIdentifier != MediaPlayerEnums::MediaEngineIdentifier::AVFoundationMSE)
+        if ((m_engineIdentifier != MediaPlayerEnums::MediaEngineIdentifier::AVFoundationMSE)
+            && (m_engineIdentifier != MediaPlayerEnums::MediaEngineIdentifier::MockMSE)) {
```
No need for two ()

jyavenard replied:
that's not what the style checker think. As soon as you have && or || on different lines , it requires all two expression to be within ()

youennf commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -6550,6 +6550,19 @@ UseGPUProcessForMediaEnabled:
+  status: embedder
+  category: media
+  humanReadableName: "GPU Process: MockMediaSource"
+  humanReadableDescription: "Run MockMediaSource in the GPU process"
```
Why do we need this option if we enable this through Internals API, which is available to WTR and API tests?
If you plan to be able to switch it on in MiniBrowser or Safari, should the status be internal?

jyavenard replied:
internal definition is `Tools for debugging the WebKit engine. Not generally useful to web developers.
#   OFF by default.` but this isn't going to be off by default for very long.

For now it's just to unable locally while I go through all the test failures and fixes them.
Once all bugs have been exposed (and fixed) I'll change the default to true.

jyavenard replied:
The internal call only enable the MockMediaSource ; it doesn't control if it is to run in the GPU Process or not; which is what that pref does.

youennf replied:
Discussed this offline and this flag should only be temporary since 'using mock' is done through internals and the MSE in GPUProcess is a different option.
Or we should rename this flag to MockMediaSourceEnabled if we want to keep it for longer and remove the internals API that allow using mock.

youennf commented on Source/WebCore/html/HTMLMediaElement.h:
```diff
@@ -495,6 +495,7 @@ class HTMLMediaElement
     WEBCORE_EXPORT static void clearMediaCache(const String&, WallTime modifiedSince = { });
     WEBCORE_EXPORT static void clearMediaCacheForOrigins(const String&, const HashSet<SecurityOriginData>&);
+
     static void resetMediaEngines();
```
Not needed

jyavenard replied:
I think it is, the comment above imply that those methods are grouped into a category. resetMediaEngines isn't part of it.

youennf commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3888,6 +3889,13 @@ void Page::abortApplePayAMSUISession(ApplePayAMSUIPaymentHandler& paymentHandler
+#if ENABLE(MEDIA_SOURCE)
+void Page::enableMockMediaSource()
+{
+    platformStrategies()->mediaStrategy().enableMockMediaSource();
```
It could be a free function or a static function.
Not sure this is worth it.

jyavenard replied:
agreed, I've removed it

youennf commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -4437,13 +4433,10 @@ bool Internals::isPluginSnapshotted(Element&)
-    WebCore::DeprecatedGlobalSettings::setGStreamerEnabled(false);
-#endif
-    MediaPlayerFactorySupport::callRegisterMediaEngine(MockMediaPlayerMediaSource::registerMediaEngine);
+    Document* document = contextDocument();
```
auto*

youennf commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -4437,13 +4433,10 @@ bool Internals::isPluginSnapshotted(Element&)
+    Document* document = contextDocument();
+    if (!document || !document->page())
+        return;
+    document->page()->enableMockMediaSource();
```
```suggestion
    platformStrategies()->mediaStrategy().enableMockMediaSource();
```

youennf commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -780,7 +782,8 @@ void MediaPlayerPrivateRemote::remoteVideoTrackConfigurationChanged(TrackPrivate
 {
-    if (m_remoteEngineIdentifier == MediaPlayerEnums::MediaEngineIdentifier::AVFoundationMSE) {
+    if ((m_remoteEngineIdentifier == MediaPlayerEnums::MediaEngineIdentifier::AVFoundationMSE)
+        || (platformStrategies()->mediaStrategy().mockMediaSourceEnabledInRemoteEngine() && m_remoteEngineIdentifier == MediaPlayerEnums::MediaEngineIdentifier::MockMSE)) {
```
OK, in the end this platformStrategies()->mediaStrategy().mockMediaSourceEnabledInRemoteEngine() should be removed, but currently it is required so that we can fix bugs after bugs.
Maybe add a FIXME to remove this check.

jyavenard replied:
that's correct. It is still of use to be able to toggle the preference to find regressions if any.

youennf commented on Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
```diff
@@ -80,4 +85,26 @@ std::unique_ptr<WebCore::NowPlayingManager> WebMediaStrategy::createNowPlayingMa
+#if ENABLE(MEDIA_SOURCE)
+void WebMediaStrategy::enableMockMediaSource()
+{
+    m_useGPUProcessForMockMediaSource = true;
```
Why are we setting it to true?
It seems then that, whenever we are using mock and GPUProcess, we will do mock in GPUProcess.
Except if after calling enableMockMediaSource(), we call setUseGPUProcessForMockMediaSource(false), which is somehow confusing.

jyavenard replied:
mistake, shouldn't have been there. it was during my test so change it without having to recompile the whole lot

youennf commented on Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:
```diff
@@ -36,6 +36,9 @@ class WebMediaStrategy : public WebCore::MediaStrategy {
 #if ENABLE(GPU_PROCESS)
     void setUseGPUProcess(bool useGPUProcess) { m_useGPUProcess = useGPUProcess; }
 #endif
+#if ENABLE(MEDIA_SOURCE)
```
class WebMediaStrategy final?

youennf commented on Source/WebCore/html/HTMLMediaElement.h:
```diff
@@ -496,7 +496,8 @@ class HTMLMediaElement
     WEBCORE_EXPORT static void clearMediaCacheForOrigins(const String&, const HashSet<SecurityOriginData>&);
-    static void resetMediaEngines();
+
+    WEBCORE_EXPORT static void resetMediaEngines();
```
Is It really necessary, cannot we just call platformStrategies()->mediaStrategy().resetMediaEngines() in the call sites?

jyavenard replied:
done

youennf reviewed (approved):
I am not sure we actually need a web preference for this. The mock and use gpuprocess web preferences might be good enough.
If this is supposed to be a temporary web preference, let's add a FIXME and file a bug to remove it.

jyavenard committed:
Run MockMediaSource in the GPU Process
https://bugs.webkit.org/show_bug.cgi?id=254074
rdar://106858091

Reviewed by Youenn Fablet.

Related changes:
- Fix crashes when running/media/media-source/media-source-minimumupcomingpresentationtime.html,
some missing methods with the RemoteSourceBufferProxy required to run this
test.
- Use C++11 initializers in MockMediaPlayerMediaSource

Covered by existing tests.

* LayoutTests/media/media-source/media-source-minimumupcomingpresentationtime.html: whitespace changes.
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::resetMediaEngines): Deleted.
* Source/WebCore/html/HTMLMediaElement.h: Delete resetMediaEngines, use mediaStragegy's resetMediaEngines instead.
* Source/WebCore/page/DeprecatedGlobalSettings.cpp: Use resetMediaEngines from mediaStrategy.
(WebCore::DeprecatedGlobalSettings::setAVFoundationEnabled):
(WebCore::DeprecatedGlobalSettings::setGStreamerEnabled):
* Source/WebCore/platform/MediaStrategy.cpp:
(WebCore::MediaStrategy::resetMediaEngines):
(WebCore::MediaStrategy::enableMockMediaSource):
(WebCore::MediaStrategy::mockMediaSourceEnabled const):
(WebCore::MediaStrategy::addMockMediaSourceEngine):
* Source/WebCore/platform/MediaStrategy.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(buildMediaEnginesVector()): Add MockMSE support to registerRemoteEngine.
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp: Use C++11 initializers.
(WebCore::MockMediaPlayerMediaSource::MockMediaPlayerMediaSource):
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::initializeMockMediaSource):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::enableMockMediaSource):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::updateCachedState):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::minimumUpcomingPresentationTimeForTrackID):
(WebKit::RemoteSourceBufferProxy::setMaximumQueueDepthForTrackID):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::load):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::minimumUpcomingPresentationTimeForTrackID):
(WebKit::SourceBufferPrivateRemote::setMaximumQueueDepthForTrackID):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::enableMockMediaSource):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:
(WebKit::WebMediaStrategy::enableMockMediaSource): Add method

Canonical link: https://commits.webkit.org/262173@main
